app/src/org/commcare/dalvik/activities/CrashWarningActivity.java

```diff
@@ -0,0 +1,136 @@
+package org.commcare.dalvik.activities;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.graphics.Color;
+import android.os.Build;
+import android.os.Bundle;
+import android.text.Spannable;
+import android.text.SpannableStringBuilder;
+import android.text.style.ForegroundColorSpan;
+import android.text.style.StyleSpan;
+import android.view.View;
+import android.widget.Button;
+import android.widget.TextView;
+
+import org.commcare.android.util.CommCareExceptionHandler;
+import org.commcare.dalvik.R;
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * Shows reason for unrecoverable crash to user and restarts CommCare
+ */
+public class CrashWarningActivity extends Activity {
+    private int errorMessageVisibility = View.GONE;
+    private static final String ERROR_VISIBLE = "error-message-is-visible";
+
+    private TextView errorMessageView;
+    private Button showErrorButton;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        loadSaveInstanceState(savedInstanceState);
+
+        setContentView(R.layout.activity_crash_warning);
+
+        setupUi();
+    }
+
+    private void loadSaveInstanceState(Bundle savedInstanceState) {
+        if (savedInstanceState != null) {
+            errorMessageVisibility =
+                    savedInstanceState.getInt(ERROR_VISIBLE, View.GONE);
+        }
+    }
+
+    private void setupUi() {
+        setupButtons();
+        setupWarningText();
+        setupErrorMessage();
+    }
+
+    private void setupButtons() {
+        Button closeButton = (Button)findViewById(R.id.RestartCommCare);
+        closeButton.setText(Localization.get("crash.warning.button"));
+        closeButton.setOnClickListener(new View.OnClickListener() {
+            public void onClick(View v) {
+                restartCommCare();
+            }
+        });
+
+        showErrorButton = (Button)findViewById(R.id.ShowError);
+        showErrorButton.setText(Localization.get("crash.show.error.button"));
+        showErrorButton.setOnClickListener(new View.OnClickListener() {
+            public void onClick(View v) {
+                toggleErrorMessageVisibility();
+            }
+        });
+    }
+
+    private void restartCommCare() {
```

Comment:
This idea of restarting CommCare is something we do in multiple places in the code, and seem to do it differently in each one. Other places off the top of my head: SingleAppManagerActivity, UnrecoverableErrorActivity. Should we try to unify this, or is there a reason that they need to be slightly different in each place? Probably need @ctsims to weigh in on this mostly.